// @liteflow/orchestrator/lib/runner.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const client_1 = require("./client");
const base58 = __importStar(require("@liteflow/api/lib/util/base58"));
class Runner extends client_1.Client {
    constructor(endpoint) {
        super(endpoint, 'Runner');
    }
    register(request, signature) {
        return __awaiter(this, void 0, void 0, function* () {
            const grpcReq = {
                envHash: base58.decode(request.envHash),
                serviceHash: base58.decode(request.serviceHash)
            };
            return this.unaryCall("Register", grpcReq, signature);
        });
    }
    delete(request, signature) {
        return __awaiter(this, void 0, void 0, function* () {
            const grpcReq = {
                runnerHash: base58.decode(request.runnerHash)
            };
            return this.unaryCall("Delete", grpcReq, signature);
        });
    }
}
exports.default = Runner;
