// @liteflow/api/lib/execution.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const lcd_1 = __importDefault(require("./util/lcd"));
var Status;
(function (Status) {
    Status[Status["Unknown"] = 0] = "Unknown";
    Status[Status["Created"] = 1] = "Created";
    Status[Status["InProgress"] = 2] = "InProgress";
    Status[Status["Completed"] = 3] = "Completed";
    Status[Status["Failed"] = 4] = "Failed";
})(Status = exports.Status || (exports.Status = {}));
class Execution extends lcd_1.default {
    get(hash) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.query(`/execution/get/${hash}`)).result;
        });
    }
    list(filter, pagination) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.query(`/execution/list`, Object.assign(Object.assign({}, filter), pagination));
            response.result.pagination = response.pagination;
            return response.result;
        });
    }
}
exports.default = Execution;
