// @liteflow/api/lib/ownership.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const lcd_1 = __importDefault(require("./util/lcd"));
var Resource;
(function (Resource) {
    Resource[Resource["None"] = 0] = "None";
    Resource[Resource["Service"] = 1] = "Service";
    Resource[Resource["Process"] = 2] = "Process";
    Resource[Resource["Runner"] = 3] = "Runner";
})(Resource = exports.Resource || (exports.Resource = {}));
class Ownership extends lcd_1.default {
    list() {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.query(`/ownership/list`)).result || [];
        });
    }
}
exports.default = Ownership;
