// @liteflow/api/lib/service.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const lcd_1 = __importDefault(require("./util/lcd"));
class Service extends lcd_1.default {
    createMsg(owner, definition) {
        const { sid, name, description, configuration, tasks, events, dependencies, repository, source } = definition;
        return {
            type: 'service/create',
            value: {
                owner,
                sid,
                name,
                description,
                configuration,
                tasks,
                events,
                dependencies,
                repository,
                source,
            }
        };
    }
    get(hash) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.query(`/service/get/${hash}`)).result;
        });
    }
    exists(hash) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.query(`/service/exist/${hash}`)).result;
        });
    }
    hash(definition) {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.query(`/service/hash`, definition, 'POST')).result;
        });
    }
    list() {
        return __awaiter(this, void 0, void 0, function* () {
            return (yield this.query('/service/list')).result || [];
        });
    }
}
exports.default = Service;
